src/Microsoft.IdentityModel.Protocols/Configuration/ConfigurationManager.cs

```diff
@@ -408,15 +429,31 @@ public override async Task<BaseConfiguration> GetBaseConfigurationAsync(Cancella
         /// <para>If <see cref="BaseConfigurationManager.RefreshInterval"/> == <see cref="TimeSpan.MaxValue"/> then this method does nothing.</para>
         /// </summary>
         public override void RequestRefresh()
+        {
+            if (AppContextSwitches.UpdateConfigAsBlocking)
+            {
+                RequestRefreshBlocking();
+            }
+            else
+            {
+                RequestRefreshBackgroundThread();
+            }
+        }
+
+        private void RequestRefreshBackgroundThread()
         {
             DateTime now = _timeProvider.GetUtcNow().UtcDateTime;
 
             if (now >= DateTimeUtil.Add(LastRequestRefresh, RefreshInterval) || _isFirstRefreshRequest)
             {
                 _isFirstRefreshRequest = false;
-                AtomicUpdateSyncAfter(now);
-                AtomicUpdateLastRequestRefresh(now);
-                _refreshRequested = true;
+
+                if (Interlocked.CompareExchange(ref _configurationRetrieverState, ConfigurationRetrieverRunning, ConfigurationRetrieverIdle) == ConfigurationRetrieverIdle)
+                {
+                    _refreshRequested = true;
+                    LastRequestRefresh = now;
+                    _updateMetadata.Set();
```

Comment:
On ln272 you are calling EnsureBackgroundTaskIsRunning before signaling. Should you do the same here?